### PR TITLE
MesonPackage: depends_on pkgconfig

### PR DESCRIPTION
**Pull request by wdconinc**
This PR adds `pkgconfig` as a dependency of all MesonPackages. At https://mesonbuild.com/Dependencies.html the use of pkg-config is front-and-center. Moreover, several MesonPackages are omitting these dependencies, e.g. https://github.com/spack/spack/pull/46912.

Currently the following MesonPackages do not list an explicit pkgconfig dependency (some of which are addressed in the PR above):
```
var/spack/repos/builtin/packages/sheenbidi/package.py
var/spack/repos/builtin/packages/toml-f/package.py
var/spack/repos/builtin/packages/libmpdclient/package.py
var/spack/repos/builtin/packages/libgudev/package.py
var/spack/repos/builtin/packages/coinhsl/package.py
var/spack/repos/builtin/packages/dpdk/package.py
var/spack/repos/builtin/packages/babl/package.py
var/spack/repos/builtin/packages/libinih/package.py
var/spack/repos/builtin/packages/libxcvt/package.py
var/spack/repos/builtin/packages/test-drive/package.py
var/spack/repos/builtin/packages/xorgproto/package.py
var/spack/repos/builtin/packages/libraqm/package.py
var/spack/repos/builtin/packages/simde/package.py
var/spack/repos/builtin/packages/glib-networking/package.py
var/spack/repos/builtin/packages/gsettings-desktop-schemas/package.py
var/spack/repos/builtin/packages/pbbam/package.py
var/spack/repos/builtin/packages/libfuse/package.py
var/spack/repos/builtin/packages/jsoncpp/package.py
var/spack/repos/builtin/packages/libmicrodns/package.py
var/spack/repos/builtin/packages/reprimand/package.py
var/spack/repos/builtin/packages/libsixel/package.py
```

### CODE REVIEW

**Comment by haampie** [comment]
Either this or a run dependency of meson itself?

File: lib/spack/spack/build_systems/meson.py [hunk 1]
```diff
@@ -46,6 +46,9 @@ class MesonPackage(spack.package_base.PackageBase):
         variant("strip", default=False, description="Strip targets on install")
         depends_on("meson", type="build")
         depends_on("ninja", type="build")
+        # Meson uses pkg-config for dependency detection, and this dependency is
+        # often overlooked by packages that use meson as a build system.
```
Comment (eli-schwartz) [inline lib/spack/spack/build_systems/meson.py:52]: > This PR adds `pkgconfig` as a dependency of all MesonPackages. At https://mesonbuild.com/Dependencies.html the use of pkg-config is front-and-center. Moreover, several MesonPackages are omitting these dependencies, e.g. #46912.
> 
> Currently the following MesonPackages do not list an explicit pkgconfig dependency (some of which are addressed in the PR above):

The question is whether it is generally the responsibility of:
- individual packages
- build systems
- some kind of global "build-essential" style selector

to depend on commonly required packages. A C/C++ compiler, GNU Make, pkg-config, they're all rather similar. And not all packages need any of that -- some packages don't even need a C compiler as they just install static data files. ;) Still, Linux distros usually treat gcc as a "build-essential" package. It's a bit different if you need to tag variant builds based on the specific compiler used, of course.

Bottom line: meson doesn't require pkg-config at all, and not all packages need it. But a whole bunch of them do, and pkg-config is fairly small, so maybe worth it to force extra dependencies for convenience?

Comment (wdconinc) [reply lib/spack/spack/build_systems/meson.py:52]: `libxcvt` is currently the only MesonPackage that does not have any dependencies and would therefore not need pkgconfig. It is possible that some other MesonPackages have dependencies that can be found without pkgconfig.

Comment (eli-schwartz) [reply lib/spack/spack/build_systems/meson.py:52]: Other MesonPackages that have dependencies will typically have to find them with either one of:
- pkg-config
- cmake

The latter handles the case where the dependency is itself built with cmake, and exports a `foobarConfig.cmake` file (cmake `find_package()`) but not a `foobar.pc` (pkg-config). Meson will create a dummy cmake project, configure it using cmake, and then read cmake's tracing API to determine which libraries the dummy project links to in the project configuration that cmake produced. Then it scrapes out those libraries and uses them as a meson dependency.

(Unfortunately, there is no other way besides literally running cmake, to determine what a foobarConfig.cmake describes.)

Comment (wdconinc) [reply lib/spack/spack/build_systems/meson.py:52]: This is interesting. So, if I read this right, then some MesonPackages maybe should also depend on `cmake`, if only to pick up the correct dependency? They may be working accidentally because `cmake` is picked up from the system rather than from spack (which means that on some other system they may fail).

Is there a way to tell meson to use only the `pkg-config` or only the `cmake` dependency resolver? In that case, we could make MesonPackage only use the methods that a specific package directly depends on. It would put the responsibility for depending on `pkg-config` again on the packages, not MesonPackage, but at least it would fail to use `pkg-config` (and probably fail to find dependencies) unless `pkg-config` was added as a dependency.

Comment (eli-schwartz) [reply lib/spack/spack/build_systems/meson.py:52]: The individual package would specify `dependency('foobar', method: 'pkgconfig')` if it knows that the foobar project only ships with .pc files. But the default is to always search first using pkg-config and if that fails, search with cmake.

Some projects ship with both pc files and cmake files, some ship with only pc files, some ship only with cmake files... it's basically a mess. And cmake is only required for the third case.

You can use meson toolchain files ("native files") to override the cmake or pkg-config binary used for dependency lookups, for example if you set it to /bin/false then the dependency lookup can't possibly work and it won't search a non-spack copy of cmake. It also respects $PKG_CONFIG as an environment variable and... I think $CMAKE also?

**Comment by wdconinc** [comment]
Ran into another package failing to build since no pkg-config. @haampie you're assigned, can you review this?

File: lib/spack/spack/build_systems/meson.py (hunk 1 above)
Comment (eli-schwartz) [reply lib/spack/spack/build_systems/meson.py:52]: What did you end up deciding about the possibility of making MesonPackage "break" pkg-config lookup when an individual package doesn't add `pkg-config` as a direct dependency?

(And maybe more importantly, what did you decide about whether to do that for *cmake*, which is an altogether less-common dependency provider?)

Comment (wdconinc) [reply lib/spack/spack/build_systems/meson.py:52]: > What did you end up deciding about the possibility of making MesonPackage "break" pkg-config lookup when an individual package doesn't add `pkg-config` as a direct dependency?

What happens in practice is that the system `pkg-config` is found, and the user doesn't see any think break while still doing the incorrect thing.

> (And maybe more importantly, what did you decide about whether to do that for _cmake_, which is an altogether less-common dependency provider?)

That's a heavier dependency which I think we should avoid. The reliance on `cmake` configs is much less common. It also introduces another dependency of one build system on another, which can lead to circular dependencies (though I don't think it would in this case).

So, summary, I think just depending on `pkg-config`, not on `cmake` is the optimum.

Comment (eli-schwartz) [reply lib/spack/spack/build_systems/meson.py:52]: The question was *specifically* about telling meson via a toolchain file that those programs are run as /bin/false, to force them to always fail.

Comment (wdconinc) [reply lib/spack/spack/build_systems/meson.py:52]: Ah, right. Hmm, gotta think how that would work best.

Is there a way to set the individual constants on the command line, without the need for a toolchain file or an environment variable? E.g. `meson setup --pkgconfig=...` instead of `meson setup --cross-file=...` or `meson setup --native-file=...`? Creating a toolchain file leads to the question of where to store it.

If we go with env vars, we could `setup_build_environment` in `meson.py`, but we don't have a way to prevent packages from overriding that.

**Comment by haampie** [comment]
I think adding pkgconfig is sensible. We can reconsider cmake in the future.

**Review by haampie — APPROVED** [review]
